### PR TITLE
S3 drivers

### PR DESCRIPTION
**Pull request by mgasvoda**

### CODE REVIEW

File: quantgov/corpora/structures.py [hunk 1]
```diff
@@ -243,3 +267,72 @@ def gen_indices_and_paths(self):
...
+        idx, path = docinfo
+        body = self.client.get_object(Bucket=self.bucket,
+                                      Key=str(path))['Body']
+        return Document(idx, body.read().decode(self.encoding))
+
+    def filter(self, pattern):
+        """ Filter paths based on index values. """
+        raise NotImplementedError
+
+    def gen_indces_and_paths(self):
```
Comment (jnelson16) [inline quantgov/corpora/structures.py]: spelling

File: quantgov/corpora/__init__.py [hunk 2]
```diff
@@ -5,5 +5,7 @@
     FlatFileCorpusDriver,
     RecursiveDirectoryCorpusDriver,
     NamePatternCorpusDriver,
-    IndexDriver
+    IndexDriver,
```
Comment (jnelson16) [inline quantgov/corpora/__init__.py]: @OliverSherouse Is there a reason this is not called `IndexCorpusDriver`, to follow the pattern above?

File: tests/test_corpora.py [hunk 3]
```diff
@@ -38,10 +38,26 @@ def build_index_corpus(directory):
...
+def build_s3_corpus(directory):
+    rows = []
+    for letter, number, path in (
+            ('a', '1', 'quantgov_tests/first.txt'),
+            ('b', '2', 'quantgov_tests/second.txt')
+    ):
+        rows.append((letter, number, path))
+    index_path = directory.join('index.csv')
+    with index_path.open('w', encoding='utf-8') as outf:
+        outf.write(u'letter,number,path\n')
```
Comment (jnelson16) [inline tests/test_corpora.py]: Should we be using the `csv.writerows` method for this? Or is this a more efficient way for testing purposes? @OliverSherouse

**Comment by jnelson16** [comment]
I have an S3Driver working with the Wyoming (or as @OliverSherouse calls it, Wisconsin) corpus!

File: quantgov/corpora/__init__.py (hunk 2 above)
Comment (OliverSherouse) [reply quantgov/corpora/__init__.py]: Nope. Let's add that as a bug and rename for 1.0

File: tests/test_corpora.py (hunk 3 above)
Comment (OliverSherouse) [reply tests/test_corpora.py]: Not more efficient, particularly, though not really a problem, either.

Comment (OliverSherouse) [inline tests/test_corpora.py]: Why do we have u strings? This ain't 2007, we're not writing python 2!

File: tests/test_corpora.py [hunk 4]
```diff
@@ -38,10 +38,26 @@ def build_index_corpus(directory):
...
+        outf.write(u'\n'.join(','.join(row) for row in rows))
+    return quantgov.corpora.S3Driver(str(index_path),
+                                     bucket='quantgov-databanks')
```
Comment (OliverSherouse) [inline tests/test_corpora.py]: Will people outside the core dev team be able to run these tests?

File: quantgov/corpora/__init__.py (hunk 2 above)
Comment (mgasvoda) [reply quantgov/corpora/__init__.py]: Added as #46

File: tests/test_corpora.py (hunk 4 above)
Comment (mgasvoda) [reply tests/test_corpora.py]: As long as they have aws credentials for boto I believe so, since the bucket is public.

**Pull request merged**